pkg/dockerregistry/server/localblobstore.go

```diff
@@ -0,0 +1,148 @@
+
+	repo *repository
+}
+
+var _ distribution.BlobStore = &localBlobStore{}
+
+func (bs *localBlobStore) Stat(ctx context.Context, dgst digest.Digest) (distribution.Descriptor, error) {
+	desc, err := bs.BlobStore.Stat(ctx, dgst)
+	switch {
+	case err == distribution.ErrBlobUnknown:
+		// continue on to the code below and look up the blob in a local store
+	case err != nil:
+		context.GetLogger(bs.repo.ctx).Errorf("Failed to find blob %s: %#v", dgst.String(), err)
+		fallthrough
+	default:
+		return desc, err
+	}
+
+	is, err := bs.repo.getImageStream()
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return distribution.Descriptor{}, distribution.ErrBlobUnknown
+		}
+		context.GetLogger(bs.repo.ctx).Errorf("Error retrieving image stream for blob: %s", err)
+
+		if errors.IsForbidden(err) {
+			return distribution.Descriptor{}, distribution.ErrBlobUnknown
+		}
+		return distribution.Descriptor{}, err
+	}
+
+	cached := bs.repo.cachedLayers.RepositoriesForDigest(dgst)
+
+	// look at the first level of tagged repositories first
+	search := bs.identifyCandidateRepositories(is, true)
+	if desc, err := bs.findCandidateRepository(ctx, search, cached, dgst); err == nil {
+		return desc, nil
+	}
+
+	// look at all other repositories tagged by the server
+	secondary := bs.identifyCandidateRepositories(is, false)
+	for k := range search {
+		delete(secondary, k)
+	}
+	if desc, err := bs.findCandidateRepository(ctx, secondary, cached, dgst); err == nil {
+		return desc, nil
+	}
+
+	return distribution.Descriptor{}, distribution.ErrBlobUnknown
+}
+
+func (bs *localBlobStore) crossStat(ctx context.Context, ref imageapi.DockerImageReference, dgst digest.Digest) (distribution.Descriptor, error) {
+	statCtx := context.WithValue(ctx, "openshift.storage.getcontent.namespace", ref.Namespace)
+	statCtx = context.WithValue(statCtx, "openshift.storage.getcontent.name", ref.Name)
+
+	desc, err := bs.BlobStore.Stat(statCtx, dgst)
+	if err != nil {
+		if err != distribution.ErrBlobUnknown {
+			context.GetLogger(bs.repo.ctx).Errorf("Unable to stat %s from %q: %v", dgst, ref.Exact(), err)
+		}
+		return distribution.Descriptor{}, err
+	}
+
+	return desc, nil
+}
+
+// TODO(legion): Merge this function with pullthroughblobstore.
+func (bs *localBlobStore) identifyCandidateRepositories(is *imageapi.ImageStream, primary bool) map[string]*imageapi.DockerImageReference {
+	search := make(map[string]*imageapi.DockerImageReference)
+	for _, tagEvent := range is.Status.Tags {
+		var candidates []imageapi.TagEvent
+		if primary {
+			if len(tagEvent.Items) == 0 {
+				continue
+			}
+			candidates = tagEvent.Items[:1]
+		} else {
+			if len(tagEvent.Items) <= 1 {
+				continue
+			}
+			candidates = tagEvent.Items[1:]
+		}
+		for _, event := range candidates {
+			ref, err := imageapi.ParseDockerImageReference(event.DockerImageReference)
+			if err != nil {
+				continue
+			}
+			if bs.repo.registryAddr != ref.Registry {
+				continue
+			}
+			ref = ref.DockerClientDefaults()
```

Comment:
I would like to merge this function with same from pullthroughblobstore into one. So I do not particularly want to they varied greatly.